File: 2015/Day1/main.cpp
#include <iostream>
#include <fstream>
#include <string>

int calculate_final_floor(const std::string &instructions)
{
    int floor = 0;
    for (char c : instructions)
    {
        if (c == '(')
            floor++;
        else if (c == ')')
            floor--;
    }
    return floor;
}

int find_basement_entry(const std::string &instructions)
{
    int floor = 0;
    for (int position = 0; position < instructions.length(); position++)
    {
        if (instructions[position] == '(')
            floor++;
        else if (instructions[position] == ')')
            floor--;

        if (floor == -1)
            return position + 1;
    }
    return -1;
}

int main()
{
    std::ifstream file("input.txt");
    if (!file.is_open())
    {
        std::cerr << "Could not open the file!" << std::endl;
        return 1;
    }

    std::string instructions;
    std::getline(file, instructions);
    file.close();

    std::cout << "Part 1: " << calculate_final_floor(instructions) << std::endl;
    std::cout << "Part 2: " << find_basement_entry(instructions) << std::endl;

    return 0;
}

File: 2015/Day1/main.lua
local function calculate_final_floor(instructions)
    local floor = 0
    for i = 1, #instructions do
        local c = instructions:sub(i, i)
        if c == '('  then
            floor = floor + 1
        elseif c == ')' then
            floor = floor - 1
        end
    end
    return floor
end

local function find_basement_entry(instructions)
    local floor = 0
    for position = 1, #instructions do
        local c = instructions:sub(position, position)
        if c == '(' then
            floor = floor + 1
        elseif c == ')' then
            floor = floor - 1
        end

        if floor == -1 then
            return position
        end
    end
    return -1
end

local file, err = io.open('input.txt', 'r')
if not file then
    print('Could not open file: ' .. err)
    os.exit(1)
end

local instructions = file:read('*all')
file:close()

print('Part 1: ' .. calculate_final_floor(instructions))
print('Part 2: ' .. find_basement_entry(instructions))
